### PR TITLE
Fix: Custom auth provider Private route redirect

### PR DESCRIPTION
**Pull request by benada002**
Like already said in https://github.com/redwoodjs/redwood/issues/2863, maybe set the loading state also in `signUp` and `reauthenticate` and maybe even `logOut`? I feel like a `loading` `true` never hurts if it gets set back to `false`. But until now I only really looked at the `AuthProvider` and `<Set private>`/`<Private>` implemantations.

### CODE REVIEW

**Comment by thedavidprice** [comment]
Thank you for moving this forward @benada002 

Looks like @dthyresson and @Tobbe are already looped in and will help you take things from here.

**Comment by Tobbe** [comment]
From a quick look, it looks alright. But I won't have time to give this a proper review, sorry :(

**Comment by benada002** [comment]
No hurry.

**Comment by thedavidprice** [comment]
@Tobbe are you able to help with this one? Or should we hand off (maybe to DT if so)?

**Comment by Tobbe** [comment]
> @Tobbe are you able to help with this one? Or should we hand off (maybe to DT if so)?

Depends on when we need it. I want to get my history PR done first. Looks like I'll have time to work on that tomorrow through the end of the week. After that I can help with this. But if we want this for v0.35 someone else will have to step in.

**Comment by thedavidprice** [comment]
Thanks @Tobbe I think it's just fine for this to come in after v0.35 (this week) and target v0.36. I just wanted to make sure it wasn't getting lost in the shuffle.

Will this timing that Tobbe suggested work out for you @benada002?

**Comment by benada002** [comment]
Yeah sure, no hurry.

**Comment by thedavidprice** [comment]
@Tobbe gentle nudge on this one.

**Comment by Tobbe** [comment]
Thanks for the nudge @thedavidprice. Might have time to look at this tomorrow

**Comment by Tobbe** [comment]
@benada002 Thanks for being so patient. 

I've taken a closer look at the code now, and agree that `loading` should be set to `true` at the beginning of all those other methods you mentioned too. But that can totally be done in a follow-up PR by you or someone else if you don't want to do it now.

It would also be very nice to have a test for this. The test would test the interaction between the auth and router packages, so I think an integration test, rather than a unit test, is the right tool for this job. In that case creating `tasks/e2e/cypress/integration/06-auth` is probably the right place for it. The other option is to treat it more like a unit test anyway, and create `packages/router/src/__tests__/auth.test.tsx` for the test. @thedavidprice what do you think?

@benada002 if you're not interested in creating a proper test for it, that's absolutely fine. I or someone else can do that. But in that case I'd ask you to just create a simple RW project that reproduces the problem that I can manually test the fix on. Let us know what you want to do.

File: packages/auth/src/AuthProvider.tsx [hunk 1]
```diff
@@ -212,6 +212,10 @@ export class AuthProvider extends React.Component<
   }
 
   logIn = async (options?: any) => {
+    this.setState({
+      ...this.state,
```
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:216]: I see that it's done this way in other places in this file, but this line really isn't needed. React will automatically do a shallow merge when updating the state.

File: packages/auth/src/AuthProvider.tsx [hunk 2]
```diff
@@ -212,6 +212,10 @@ export class AuthProvider extends React.Component<
...
+      loading: true,
+    })
```
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:218]:
```suggestion
    this.setState({ loading: true })
```

**Comment by benada002** [comment]
I run while testing the test project using `rwt link` into an error (normal Redwood works fine). I'll try to send a screenshot tomorrow.

**Comment by thedavidprice** [comment]
CI error was a network issue with Windows. Just re-running... expect everything to pass

**Comment by thedavidprice** [comment]
@benada002 I wouldn't worry too much about `rwt link` having issues. The CI is passing, which is great. Also, we are soon to replace the `link` option with a new command for linking/syncing. 

From you project, try running `yarn rwfw project:sync` You'll need to pass the path to your Framework if you haven't set it up as an Env Var ([see how in this doc](https://github.com/redwoodjs/redwood/blob/main/CONTRIBUTING.md#specifying-a-rw_path)). Apologies that we haven't updated the Contributing doc yet. It's coming!

**Comment by benada002** [comment]
@Tobbe I'd be interested to write the tests, but I've to dig into cypress first. So, here comes the repro first: https://github.com/benada002/redwood-repro-2872.

**Comment by benada002** [comment]
@thedavidprice `yarn rwfw project:sync` was working like a charm, thanks.

**Comment by Tobbe** [comment]
I'll get the repro up and running locally tonight. If everything looks good we'll merge this and then we can follow up with another PR with automated tests later

**Comment by Tobbe** [comment]
@benada002 Thanks for the great repro. Made it super easy to test 👍

**Comment by thedavidprice** [comment]
Great work on this @benada002 🚀

**Pull request merged**